File: g5-admin/src/api/client/master-lock.ts
import type { MasterLockSetupInput } from "../../types/MasterLockSetupInput";
import type { MasterLockStatus } from "../../types/MasterLockStatus";
import type { MasterLockTotpInput } from "../../types/MasterLockTotpInput";
import type { MasterLockUnlockInput } from "../../types/MasterLockUnlockInput";
import { invokeCommand } from "./core";

export async function getMasterLockStatus(): Promise<MasterLockStatus> {
  return invokeCommand<MasterLockStatus>("cmd_master_lock_status");
}

export async function setupMasterLock(
  input: MasterLockSetupInput,
): Promise<MasterLockStatus> {
  return invokeCommand<MasterLockStatus>("cmd_master_lock_setup", { input });
}

export async function unlockMasterLock(
  input: MasterLockUnlockInput,
): Promise<MasterLockStatus> {
  return invokeCommand<MasterLockStatus>("cmd_master_lock_unlock", { input });
}

export async function unlockFastMasterLock(): Promise<MasterLockStatus> {
  return invokeCommand<MasterLockStatus>("cmd_master_lock_unlock_fast");
}

export async function lockMasterLock(): Promise<MasterLockStatus> {
  return invokeCommand<MasterLockStatus>("cmd_master_lock_lock");
}

export async function verifyMasterLockTotp(
  input: MasterLockTotpInput,
): Promise<MasterLockStatus> {
  return invokeCommand<MasterLockStatus>("cmd_master_lock_verify_totp", { input });
}

File: g5-admin/src-tauri/src/commands/master_lock.rs
use crate::app_state::AppState;
use crate::commands::common::command_error_payload;
use crate::error::CommandResult;
use crate::fast_unlock;
use crate::request_id::next_request_id;
use g5_admin_models::models::master_lock::{
    MasterLockSetupInput, MasterLockStatus, MasterLockTotpInput, MasterLockUnlockInput,
};
use tauri::{AppHandle, Runtime, State};

const MASTER_LOCK_COMPONENT: &str = "g5_admin::commands::master_lock";

#[tauri::command]
pub async fn cmd_master_lock_status(state: State<'_, AppState>) -> CommandResult<MasterLockStatus> {
    let request_id = next_request_id();
    state
        .master_lock_service()
        .master_lock_status(&request_id)
        .await
        .map_err(|error| {
            command_error_payload(
                MASTER_LOCK_COMPONENT,
                "cmd_master_lock_status",
                "local-master-lock",
                &request_id,
                error,
            )
        })
}

#[tauri::command]
pub async fn cmd_master_lock_setup(
    state: State<'_, AppState>,
    input: MasterLockSetupInput,
) -> CommandResult<MasterLockStatus> {
    let request_id = next_request_id();
    state
        .master_lock_service()
        .setup_master_lock(&request_id, input)
        .await
        .map_err(|error| {
            command_error_payload(
                MASTER_LOCK_COMPONENT,
                "cmd_master_lock_setup",
                "local-master-lock",
                &request_id,
                error,
            )
        })
}

#[tauri::command]
pub async fn cmd_master_lock_unlock(
    state: State<'_, AppState>,
    input: MasterLockUnlockInput,
) -> CommandResult<MasterLockStatus> {
    let request_id = next_request_id();
    state
        .master_lock_service()
        .unlock_master_lock(&request_id, input)
        .await
        .map_err(|error| {
            command_error_payload(
                MASTER_LOCK_COMPONENT,
                "cmd_master_lock_unlock",
                "local-master-lock",
                &request_id,
                error,
            )
        })
}

#[tauri::command]
pub async fn cmd_master_lock_verify_totp(
    state: State<'_, AppState>,
    input: MasterLockTotpInput,
) -> CommandResult<MasterLockStatus> {
    let request_id = next_request_id();
    state
        .master_lock_service()
        .verify_master_lock_totp(&request_id, input)
        .await
        .map_err(|error| {
            command_error_payload(
                MASTER_LOCK_COMPONENT,
                "cmd_master_lock_verify_totp",
                "local-master-lock",
                &request_id,
                error,
            )
        })
}

#[tauri::command]
pub async fn cmd_master_lock_unlock_fast<R: Runtime>(
    state: State<'_, AppState>,
    app: AppHandle<R>,
) -> CommandResult<MasterLockStatus> {
    let request_id = next_request_id();
    let secret = fast_unlock::load_secret(&app).map_err(|error| {
        command_error_payload(
            MASTER_LOCK_COMPONENT,
            "cmd_master_lock_unlock_fast",
            "local-master-lock",
            &request_id,
            error,
        )
    })?;

    state
        .master_lock_service()
        .unlock_master_lock_fast(&request_id, &secret)
        .await
        .map_err(|error| {
            command_error_payload(
                MASTER_LOCK_COMPONENT,
                "cmd_master_lock_unlock_fast",
                "local-master-lock",
                &request_id,
                error,
            )
        })
}

#[tauri::command]
pub async fn cmd_master_lock_lock(state: State<'_, AppState>) -> CommandResult<MasterLockStatus> {
    let request_id = next_request_id();
    state
        .master_lock_service()
        .lock_master(&request_id)
        .await
        .map_err(|error| {
            command_error_payload(
                MASTER_LOCK_COMPONENT,
                "cmd_master_lock_lock",
                "local-master-lock",
                &request_id,
                error,
            )
        })
}
